### ISSUE TITLE
Outdated alc.exe in latest CU images

### ISSUE THREAD

**Issue by ShadowXVII** [microsoft/navcontainerhelper#903]
The images appear to contain an older version of `alc.exe`
Is there a way to update this, or would it be updated in the images online?

i.e. `onprem:14.9.39327.0-au-ltsc2019` contains `alc.exe` version `3.0.2.62814.`

We've an app that has a compile bug (well, strictly a style bug) where it incorrectly flags a line as needing a prefix. This error doesn't occur on alc.exe version `4.0.3.33113`.

We can obviously work around the rule by disabling the checks in the json. Or, compiling for a later version, but we need to backwards support older versions (i.e. v140). 

The hosted/online images also contain the old version.

**Comment by freddydk** [comment]
I would assume that 14.9 comes with the AL compiler from 14.9, which would be the same you would find if you download the 14.9 DVD and install that.
Are you telling me that this is not the case?

**Comment by ShadowXVII** [comment]
Hi

Well, I suspect it does, because that's the alc it's try to use to compile. Which doesn't match the compiler in the latest vs code extension.

My issue is Microsoft don't seem to be bundling the latest compiler in the cumulative updates?

If an issue is fixed in vX of the compiler, shouldn't it be backported to BC 14.X? It's a bug fix.

Regards
Jake

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by freddydk** [comment]
Typically, partners can request a hotfix to be backported, I don't know if that is the case for compiler fixes, you would have to ask CSS about this.
What you can do with containers is to create the container with the -vsixFile parameter and specify a URL or file path to a .vsix file (you can download the .vsix from the marketplace).

**Comment by ShadowXVII** [comment]
Hi Freddy

Ahk, yes, I'll have to do that. Any chance we can get alc.exe docker images?

Otherwise I'll need to store the visx somewhere, and updating it is a pain. Maybe I can PowerShell download the extension from the vs code store and extract...

Regards
Jake

**Comment by freddydk** [comment]
I was trying to find a way to get the "latest" from marketplace - didn't succeed.
The best way to programmatically get the latest would be something like this:
```
$imageName = "mcr.microsoft.com/businesscentral/onprem:w1-ltsc2019"
Remove-Item -Path "C:\ProgramData\NavContainerHelper\latestvsix" -Recurse -Force
docker pull $imageName
Extract-FilesFromBCContainerImage -imageName $imageName -path "C:\ProgramData\NavContainerHelper\latestvsix" -extract vsix
$vsixFile = (Get-Item -path "C:\ProgramData\NavContainerHelper\latestvsix\*.vsix").FullName
$vsixFile
```

**Comment by ShadowXVII** [comment]
I did some digging and put together this option -- not sure how fragile it is though.

```
$listing = Invoke-WebRequest -Method POST  `
                  -Uri https://marketplace.visualstudio.com/_apis/public/gallery/extensionquery?api-version=3.0-preview.1 `
                  -Body '{"filters":[{"criteria":[{"filterType":8,"value":"Microsoft.VisualStudio.Code"},{"filterType":12,"value":"4096"},{"filterType":7,"value":"ms-dynamics-smb.al"}],"pageNumber":1,"pageSize":50,"sortBy":0,"sortOrder":0}],"assetTypes":[],"flags":914}' `
                  -ContentType application/json | ConvertFrom-Json 

// $listing.$results[0].extensions[0].versions[0].files
$visxUrl = $listing.results | Select-Object -First 1 -ExpandProperty extensions `
                     | Select-Object -First 1 -ExpandProperty versions `
                     | Select-Object -First 1 -ExpandProperty files `
                     | Where-Object { $_.assetType -eq "Microsoft.VisualStudio.Services.VSIXPackage"} `
                     | Select-Object -ExpandProperty source

Write-Host "Found url '$visxUrl'"
$listing = Invoke-WebRequest -Method GET -Uri $visxUrl -OutFile 'ms-dynamics-smb.al.visx'
```

**Comment by freddydk** [comment]
Thanks - will check with the R&D team whether this is useful.
The URL can directly be used as a parameter to -vsixFile - you don't have to download first.

**Comment by freddydk** [comment]
This looks very useful. Will consider adding a latest option (magic) to new-navcontainer, which does exactly this. Thanks a million.

**Comment by ShadowXVII** [comment]
Ooh yay, -MagicVisx ... or I'll settle for -GetLatestVisx 😂

**Comment by freddydk** [comment]
Maybe just a new function called Get-LatestVsixUrl (your code) and then -vsixFile (Get-LatestVsixUrl)

**Comment by ShadowXVII** [comment]
Sweet, I've implemented this in my build pipeline.

However, the NAV container helper assumes it's a *.visx file when it's in-fact a "Microsoft.VisualStudio.Services.VSIXPackage" file. Same difference, just if I'm using the "auto download" via `vsixFile` it doesn't work.

**Comment by freddydk** [comment]
Will ship in 0.6.5.1 (very soon)

**Comment by freddydk** [comment]
Fixed in 0.6.5.1 and 0.6.5.2
